Bug fixes in MYNN surface layer scheme. Also, added lat/log to RUC LSM for debugging.

tanyasmirnova opened the pull request:
1. Fixed bugs in MYNN surface layer scheme in land/ice/water parts of the code
    to avoid using in computations "huge" numbers of qsfc_lnd, qsfc_ice and qsfc_wat.
2. Fixed bugs in MYNN surface layer scheme for initialization and consistent
    use of surface QV from RUC LSM over land and ice.
3. Passed into RUC LSM driver lat/lon for debug prints for a test point with known lat/lon.

tanyasmirnova committed:
Pass into the RUC driver lat/lons for debugging.

tanyasmirnova committed:
Fixed bugs in MYNN surface layer scheme in land/ice/water parts of the code
to avoid using in computations "huge" numbers of qsfc_lnd, qsfc_ice and qsfc_wat.
Also, fixed bugs in MYNN surface layer scheme for initializattion and consistent
use of surface QV from RUC LSM over land and ice.

joeolson42 commented on physics/module_sf_mynn.F90:
```diff
@@ -659,12 +659,73 @@ SUBROUTINE SFCLAY1D_mynn(                                      &
            write(0,*)"=== important input to mynnsfclayer, i:", i
            IF (dry(i)) THEN
-             write(0,*)"dry=",dry(i)," pblh=",pblh(i)," tsk=", tskin_lnd(i),&
+             write(0,*)"dry=",dry(i)," =",pblh(i)," tsk=", tskin_lnd(i),&
```
Lets keep the "pblh"

tanyasmirnova replied:
@joeolson42 Sorry, Joe, I removed it by accident. Just committed the fix to that.

joeolson42 commented on physics/module_sf_mynn.F90:
```diff
@@ -728,6 +796,7 @@ SUBROUTINE SFCLAY1D_mynn(                                      &
          GOVRTH(I)=G/TH1D(I)
       ENDDO
 
+      !tgs - should QFX and HFX be separate for land, ice and water?
```
If we go back to using the flux calculations, then yes they should be broken down into components.

joeolson42 commented on physics/module_sf_mynn.F90:
```diff
@@ -1616,6 +1636,7 @@ SUBROUTINE SFCLAY1D_mynn(                                      &
 
        ! Compute u* without vconv for use in HFX calc when isftcflx > 0           
        WSPDI(I)=MAX(SQRT(U1D(I)*U1D(I)+V1D(I)*V1D(I)), wmin)
+       !tgs - should USTM be separater for dry, icy, wet?
```
USTM was a variable put in the scheme by the hurricane community a long time ago. I'm not sure if anyone uses it.

joeolson42 commented on physics/module_sf_mynn.F90:
```diff
@@ -1662,6 +1686,7 @@ SUBROUTINE SFCLAY1D_mynn(                                      &
     !----AND COMPUTE THE MOISTURE SCALE (or q*)
     !----------------------------------------------------
 
+    !tgs - should we have MOL and qstar separate for dry, icy and wet?
```
We can probably make use out of both - especially for the negative qv problem. I think having a separate heat and moisture transfer coefficients may be the only way to fix that problem.

joeolson42 commented on physics/module_sf_mynn.F90:
```diff
@@ -1750,15 +1775,16 @@ SUBROUTINE SFCLAY1D_mynn(                                      &
          ! CALCULATE THE EXCHANGE COEFFICIENTS FOR HEAT (FLHC)
          ! AND MOISTURE (FLQC)
          !------------------------------------------
+         !tgs - should FLQC, FLHC be separate for dry, icy and wet?
```
All of the rest of the questions are up to us. If we go back to using the fluxes, which I think we want to do, then yes, we should break everything up into different types.

tanyasmirnova replied:
I think we should go back to MYNN surface fluxes over water.

joeolson42 reviewed (approved):
All good, except for deleting the "pblh" in the print statement in module_sf_mynn.F90.

tanyasmirnova committed:
Restored modified by accident print statement.

DomHeinzeller reviewed (commented):
Looks great to the extent that I can tell. Let me run the regression tests with it (and create the fv3atm and ufs-weather-model PRs).

DomHeinzeller commented on physics/sfc_drv_ruc.meta:
```diff
@@ -729,6 +729,24 @@
+  standard_name = longitude_in_degree
+  long_name = longitude in degree east
+  units = degree_east
+  dimensions = (horizontal_dimension)
```
@tanyasmirnova all the debug runs in `rt_ccpp_dev.conf` that use RUC LSM crashed with
```
138: An error occurred in ccpp_physics_run for group physics, block    1 and thread    1 (ntX=   1)
138: Detected size mismatch for variable GFS_Data(cdata%blk_no)%Grid%xlon_d in group physics before lsm_ruc_run, expected      384 but got       32
```
I think this is because line 745 says `(horizontal_dimension)` instead of `(horizontal_loop_extent)`.

Can you please fix this and run at least one manual test in DEBUG mode?

Thanks!

tanyasmirnova committed:
Fixed the bug in horizontal dimension for xlon_d.

climbfuji commented:
Thanks, Tanya! I kicked off the tests again.

tanyasmirnova commented:
@climbfuji Thank you, Dom! I tried to run my usual C768 test in the debug mode, but it crashed with a different error:
....
 652: WARNING from PE   652: diag_manager_mod::send_data_3d: A value for dynamics in field slp (Min:   1.01429E+003, Max:   1.20970E+003) is outside the range [  8.00000E+002,  1.20000E+003], and not equal to the missing value.
 652:
 507: forrtl: error (65): floating invalid
 507: Image              PC                Routine            Line        Source
 507: fv3.exe            000000000B0B948E  Unknown               Unknown  Unknown
 507: libpthread-2.17.s  00002B9644110630  Unknown               Unknown  Unknown
 507: fv3.exe            000000000611C59B  Unknown               Unknown  Unknown
 507: fv3.exe            0000000003E2B491  fv_mapz_mod_mp_la         556  fv_mapz.F90

Maybe it was caused by the FMS problem in the top of the gsl/develop branch  that we saw before this PR?

climbfuji commented:
> @climbfuji Thank you, Dom! I tried to run my usual C768 test in the debug mode, but it crashed with a different error: .... 652: WARNING from PE 652: diag_manager_mod::send_data_3d: A value for dynamics in field slp (Min: 1.01429E+003, Max: 1.20970E+003) is outside the range [ 8.00000E+002, 1.20000E+003], and not equal to the missing value. 652: 507: forrtl: error (65): floating invalid 507: Image PC Routine Line Source 507: fv3.exe 000000000B0B948E Unknown Unknown Unknown 507: libpthread-2.17.s 00002B9644110630 Unknown Unknown Unknown 507: fv3.exe 000000000611C59B Unknown Unknown Unknown 507: fv3.exe 0000000003E2B491 fv_mapz_mod_mp_la 556 fv_mapz.F90
> 
> Maybe it was caused by the FMS problem in the top of the gsl/develop branch that we saw before this PR?

Probably.

Even with your bug fix, I still get crashes for all debug tests. I can't see why (hera is down today), so I will need to rerun the tests on Cheyenne. Stay tuned.

DomHeinzeller commented:
There is still a problem with `tsk_wat` in MYNN SFC it seems. See attached screenshot.

![Screen Shot 2021-11-02 at 11 03 52 AM](https://user-images.githubusercontent.com/58610420/139911877-85dcd8fa-42e4-4d89-92ea-44b21d84c39d.png)

DomHeinzeller commented:
@tanyasmirnova I think that in lines 664 to 746 in `module_sf_mynn.F90` we should be using `tskin_*` instead of the local and yet uninitialized variables `tsk_*`. Can you confirm?

tanyasmirnova commented:
@climbfuji Dom, I have a fix for this error. I will commit it soon. Please, stand by.

joeolson42 commented:
Dom,
I think you are right. The TSK_* variables are not defined until lines
755-771.
-joe

On Tue, Nov 2, 2021 at 11:36 AM DomHeinzeller ***@***.***>
wrote:

> @tanyasmirnova <https://github.com/tanyasmirnova> I think that in lines
> 664 to 746 in module_sf_mynn.F90 we should be using tskin_* instead of
> the local and yet uninitialized variables tsk_*. Can you confirm?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSL/ccpp-physics/pull/112#issuecomment-957961242>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADLRR3U6YGPTENOQTST5KPLUKATKDANCNFSM5HCUCOPA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Joseph Olson
Model Physics Branch Chief
Environmental Prediction Advancement Division
NOAA-Global Systems Laboratory
Boulder, Colorado

tanyasmirnova commented:
Joe, yes, this is the problem. I moved this piece of code to the beginning, and TSK_* are not defined yet. Should TSK_wat be 0.5 *(tskin_wat+tsurf_wat) even at ITIMESTEP=1?

tanyasmirnova committed:
Added computations of local TSK_wat, TSK_ice and TSK_lnd at ITIMESTEP=1
before they get used.

joeolson42 commented:
I just changed the first few instances of TSK_* to TSKIN_* but it still
crashed in sfc_drv_ruc.F90:

forrtl: severe (408): fort: (3): Subscript #1 of the array LEMITBL has
value 0 which is less than the lower bound of 1


Image              PC                Routine            Line        Source

fv3_gfs.x          0000000008B01750  Unknown               Unknown  Unknown

fv3_gfs.x          00000000079D52CB  lsm_ruc_mp_lsm_ru         170
sfc_drv_ruc.F90

fv3_gfs.x          0000000006D6F1B8  ccpp_fv3_hrrr_phy         617
ccpp_FV3_HRRR_physics_cap.F90

fv3_gfs.x          0000000006D57CF1  ccpp_fv3_hrrr_cap         157
ccpp_FV3_HRRR_cap.F90

fv3_gfs.x          00000000069064C5  ccpp_static_api_m         329
ccpp_static_api.F90

fv3_gfs.x          00000000068FC834  ccpp_driver_mp_cc         108
CCPP_driver.F90

fv3_gfs.x          00000000033B3D3C  atmos_model_mod_m         722
atmos_model.F90

fv3_gfs.x          0000000002D453C9  module_fcst_grid_         417
module_fcst_grid_comp.F90




On Tue, Nov 2, 2021 at 12:11 PM tanyasmirnova ***@***.***>
wrote:

> Joe, yes, this is the problem. I moved this piece of code to the
> beginning, and TSK_* are not defined yet. Should TSK_wat be 0.5
> *(tskin_wat+tsurf_wat) even at ITIMESTE=1?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSL/ccpp-physics/pull/112#issuecomment-957986285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADLRR3WCF4XTK345RPPZRXTUKAW7VANCNFSM5HCUCOPA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Joseph Olson
Model Physics Branch Chief
Environmental Prediction Advancement Division
NOAA-Global Systems Laboratory
Boulder, Colorado

DomHeinzeller commented:
> Joe, yes, this is the problem. I moved this piece of code to the beginning, and TSK_* are not defined yet. Should TSK_wat be 0.5 *(tskin_wat+tsurf_wat) even at ITIMESTEP=1?



> I just changed the first few instances of TSK_* to TSKIN_* but it still crashed in sfc_drv_ruc.F90: forrtl: severe (408): fort: (3): Subscript #1 of the array LEMITBL has value 0 which is less than the lower bound of 1 Image PC Routine Line Source fv3_gfs.x 0000000008B01750 Unknown Unknown Unknown fv3_gfs.x 00000000079D52CB lsm_ruc_mp_lsm_ru 170 sfc_drv_ruc.F90 fv3_gfs.x 0000000006D6F1B8 ccpp_fv3_hrrr_phy 617 ccpp_FV3_HRRR_physics_cap.F90 fv3_gfs.x 0000000006D57CF1 ccpp_fv3_hrrr_cap 157 ccpp_FV3_HRRR_cap.F90 fv3_gfs.x 00000000069064C5 ccpp_static_api_m 329 ccpp_static_api.F90 fv3_gfs.x 00000000068FC834 ccpp_driver_mp_cc 108 CCPP_driver.F90 fv3_gfs.x 00000000033B3D3C atmos_model_mod_m 722 atmos_model.F90 fv3_gfs.x 0000000002D453C9 module_fcst_grid_ 417 module_fcst_grid_comp.F90
> […](#)
> On Tue, Nov 2, 2021 at 12:11 PM tanyasmirnova ***@***.***> wrote: Joe, yes, this is the problem. I moved this piece of code to the beginning, and TSK_* are not defined yet. Should TSK_wat be 0.5 *(tskin_wat+tsurf_wat) even at ITIMESTE=1? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#112 (comment)](https://github.com/NOAA-GSL/ccpp-physics/pull/112#issuecomment-957986285)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADLRR3WCF4XTK345RPPZRXTUKAW7VANCNFSM5HCUCOPA> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
> -- Joseph Olson Model Physics Branch Chief Environmental Prediction Advancement Division NOAA-Global Systems Laboratory Boulder, Colorado

The regression tests are still crashing in MYNN SFC. Looking at it.

DomHeinzeller commented:
@tanyasmirnova @joeolson42 First of all, I had to make this change:
```
diff --git a/physics/module_sf_mynn.F90 b/physics/module_sf_mynn.F90
index 1ddee4bc..41e41991 100644
--- a/physics/module_sf_mynn.F90
+++ b/physics/module_sf_mynn.F90
@@ -661,6 +661,9 @@ CONTAINS
 !-------------------------------------------------------------------
       DO I=its,ite

+         ! PSFC ( in cmb) is used later in saturation checks
+         PSFC(I)=PSFCPA(I)/1000.
+
          IF (ITIMESTEP == 1) THEN
          !initialize surface specific humidity and mixing ratios for land, ice and water
             IF (wet(i)) THEN
@@ -749,8 +752,6 @@ CONTAINS
       ENDIF

       DO I=its,ite
-         ! PSFC ( in cmb) is used later in saturation checks
-         PSFC(I)=PSFCPA(I)/1000.
          QVSH(I)=QV1D(I)/(1.+QV1D(I))        !CONVERT TO SPEC HUM (kg/kg)
          THCON(I)=(100000./PSFCPA(I))**ROVCP
          ! DEFINE SKIN TEMPERATURES FOR LAND/WATER/ICE
```
Now the code is crashing in in `module_sf_mynn.F90` in line 1181, because `ZT_lnd(i)` is zero:
```
       GZ1OZt_lnd(I)= LOG((ZA(I)+ZNTstoch_lnd(i))/ZT_lnd(i))
```
Can this be fixed?

tanyasmirnova commented:
@joeolson42 @climbfuji Dom and Joe, I do not have any of the problems that you see in my test case. I run FV3_HRRR suite in the 3-km NA RRFS test case.
I put a print statement for ZT_lnd, and it never had a zero value in my test.

climbfuji commented:
The `fv3_gsd_noah_mynnsfc_debug` regression test has the problem (and I think the other mynnsfc debug regression tests, too). They are using a binary landmask, you are using a fractional mask, correct? Maybe that makes the difference?

tanyasmirnova commented:
Dom, we use frac_grid=.false. in RRFS. My guess is that MYNN surface layer might have problems with frac_grid=.true.

joeolson42 commented:
I don't see any values of 0 for zt*, but I am still getting a crash in sfc_drv_ruc.F90:

276: forrtl: severe (408): fort: (3): Subscript #1 of the array LEMITBL has value 0 which is less than the lower bound of 1
276:
276: Image              PC                Routine            Line        Source
276: fv3_gfs.x          0000000008B08B90  Unknown               Unknown  Unknown
276: fv3_gfs.x          00000000079D52D7  lsm_ruc_mp_lsm_ru         170  sfc_drv_ruc.F90
276: fv3_gfs.x          0000000006D6F1B8  ccpp_fv3_hrrr_phy         617  ccpp_FV3_HRRR_physics_cap.F90
276: fv3_gfs.x          0000000006D57CF1  ccpp_fv3_hrrr_cap         157  ccpp_FV3_HRRR_cap.F90
276: fv3_gfs.x          00000000069064C5  ccpp_static_api_m         329  ccpp_static_api.F90

tanyasmirnova commented:
Joe,
The following line:
semisbase(i) = lemitbl(vtype(i)) 
just uses the table value of emissivity for the veg. type. Maybe you do not have Dom's changes to move vtype from real to integer?

DomHeinzeller commented:
So far I tracked it back to ustar and restart having large values, which then cause the equations
```
          Zt = Z_0*EXP(-KARMAN*CZIL*SQRT(restar))
          Zt = MIN( Zt, 0.75*Z_0)

          Zq = Z_0*EXP(-KARMAN*CZIL*SQRT(restar))
          Zq = MIN( Zq, 0.75*Z_0)
```
to produce zeros: ustar and restar are something like `9113164117.8289948`

joeolson42 commented:
There must be an instability somewhere further upstream. The frictional
velocity and the Reynolds number should never get that large.

I need to update my code. Does this crash only happen with the fractional
grid option? I

On Tue, Nov 2, 2021 at 4:21 PM DomHeinzeller ***@***.***>
wrote:

> So far I tracked it back to ustar and restart having large values, which
> then cause the equations
>
>           Zt = Z_0*EXP(-KARMAN*CZIL*SQRT(restar))
>           Zt = MIN( Zt, 0.75*Z_0)
>
>           Zq = Z_0*EXP(-KARMAN*CZIL*SQRT(restar))
>           Zq = MIN( Zq, 0.75*Z_0)
>
> to produce zeros: ustar and restar are something like 9113164117.8289948
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSL/ccpp-physics/pull/112#issuecomment-958271012>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADLRR3S2U2GGURXE6MEXXCTUKBW7FANCNFSM5HCUCOPA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Joseph Olson
Model Physics Branch Chief
Environmental Prediction Advancement Division
NOAA-Global Systems Laboratory
Boulder, Colorado

DomHeinzeller commented:
No, this is an out of the box regression test C96/l64/non-frac that runs fine with the head of gsl/develop. The problem must be in this PR.

> On Nov 2, 2021, at 4:50 PM, Joseph Olson ***@***.***> wrote:
> 
> 
> There must be an instability somewhere further upstream. The frictional
> velocity and the Reynolds number should never get that large.
> 
> I need to update my code. Does this crash only happen with the fractional
> grid option? I
> 
> On Tue, Nov 2, 2021 at 4:21 PM DomHeinzeller ***@***.***>
> wrote:
> 
> > So far I tracked it back to ustar and restart having large values, which
> > then cause the equations
> >
> > Zt = Z_0*EXP(-KARMAN*CZIL*SQRT(restar))
> > Zt = MIN( Zt, 0.75*Z_0)
> >
> > Zq = Z_0*EXP(-KARMAN*CZIL*SQRT(restar))
> > Zq = MIN( Zq, 0.75*Z_0)
> >
> > to produce zeros: ustar and restar are something like 9113164117.8289948
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/NOAA-GSL/ccpp-physics/pull/112#issuecomment-958271012>,
> > or unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/ADLRR3S2U2GGURXE6MEXXCTUKBW7FANCNFSM5HCUCOPA>
> > .
> > Triage notifications on the go with GitHub Mobile for iOS
> > <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> > or Android
> > <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
> >
> >
> 
> 
> -- 
> Joseph Olson
> Model Physics Branch Chief
> Environmental Prediction Advancement Division
> NOAA-Global Systems Laboratory
> Boulder, Colorado
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/NOAA-GSL/ccpp-physics/pull/112#issuecomment-958328758>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AN7FF5DAR2QK777G55JYJ2TUKBTKDANCNFSM5HCUCOPA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>

tanyasmirnova commented:
I have changed the following lines in this PR:
      DO I=its,ite
         ! CONVERT SKIN POTENTIAL TEMPERATURES TO VIRTUAL POTENTIAL TEMPERATURE:
         THVSK_lnd(I) = THSK_lnd(I)*(1.+EP1*QVSH(I))   !(K)
         THVSK_ice(I) = THSK_ice(I)*(1.+EP1*QVSH(I))   !(K)
         THVSK_wat(I) = THSK_wat(I)*(1.+EP1*QVSH(I))   !(K)
      ENDDO
Because Huge values were used in these lines. Iput THVSK computation for land into "dry" section, and so on. I also replaced QVSH (computed from the 1st level QV) with qsfc_lnd for land. I had problems with these lines, but got it work fine with RUC LSM. I added the check on THVSK:
           if(THVSK_lnd(I) < 170. .or. THVSK_lnd(I) > 350.) &
           print *,'THVSK_lnd(I)',i,THVSK_lnd(I),THSK_lnd(i),tsurf_lnd(i),tskin_lnd(i),qsfc_lnd(i)

Dom, do you see any prints for THVSK_lnd?

tanyasmirnova committed:
Move PSFC computation to the beginning before it is used.

tanyasmirnova committed:
1. Pass in flag_iter to MYNNSFC_wrapper for consistency with iterations in LSM, ice and
ocean. ! when iter = 1, flag_iter = .true. for all grids ! when iter =
2, flag_iter = .true. when wind < 2 for both land and ocean (when
nstf_name1 > 0)
2.Added flag_iter to all computations in MYNN surface layer scheme to
avoid inconsistencies with iterations in LSM, ice andocean models.
3.Removed averaging: 0.5*(tsurf + tskin) and used tskin instead, because
tsurf could be not defined.
4.Added checks on QSFC for non-defined values in LSM, ice and ocean models.
If there are not defined values, compute QSFC as at itimestep=1.

tanyasmirnova commented:
I think I resolved the issues with the Noah LSM. I ran my C&^* test case in the debug mode, and it runs fine now.
Dom, could you, please, try again the regression tests? Thank you!

climbfuji commented:
> I think I resolved the issues with the Noah LSM. I ran my C&^* test case in the debug mode, and it runs fine now.
> Dom, could you, please, try again the regression tests? Thank you!

Running this code now ...

tanyasmirnova commented:
@climbfuji Dom, thank you very much for your help with this PR. Greatly appreciated!
